src/uct/ib/rc/verbs/rc_verbs_iface.c

```diff
@@ -109,17 +109,34 @@ uct_rc_verbs_iface_poll_tx(uct_rc_verbs_iface_t *iface)
         count = uct_rc_verbs_txcq_get_comp_count(&wc[i], &ep->super.txqp);
         ucs_trace_poll("rc_verbs iface %p tx_wc wrid 0x%lx ep %p qpn 0x%x count %d",
                        iface, wc[i].wr_id, ep, wc[i].qp_num, count);
-        uct_rc_verbs_txqp_completed(&ep->super.txqp, &ep->txcnt, count);
+        ep->txcnt.ci += count;
+
+        /* Process completions before updating resources to avoid OOO.
+         * Some resources (RDMA_READ credits) are updated in operation completion
+         * callbacks, so need to call all relevant callbacks prior dispatching
+         * arbiter. In this case ordering is preserved, because:
+         * - If nothing is sent from the completion callback, we just update
+         *   resources and dispatch arbiter.
+         * - If something needs to be sent from completion callback, but not all
+         *   needed resources available, it will be added to the pending queue
+         *   (and handled in dispatch below in the proper order).
+         */
```

Comment:
a comment is duplicated, can we somehow wrap mlx5 and verbs code to the wrapper that will have this comment?

Reply:
not sure it will be better, this comment is intended to be in `poll_tx` itself to warn that sequence of funcs is important.

Reply:
I see
but it is hard to maintain 3 equal comments...